rclc/src/rclc/action_goal_handle.c

```diff
@@ -0,0 +1,283 @@
+  while (NULL != handle) {
+    if (handle == goal_handle) {
+      return true;
+    }
+    handle = handle->next;
+  }
+  return false;
+}
+
+rclc_action_goal_handle_t * rclc_action_pop_goal_handle_from_list(
+  rclc_action_goal_handle_t ** list)
+{
+  rclc_action_goal_handle_t * handle = *list;
+  *list = (*list == NULL ) ? NULL : (*list)->next;
+  return handle;
+}
+
+rclc_action_goal_handle_t * rclc_action_take_goal_handle_from_list(
+  rclc_action_goal_handle_t ** list,
+  rclc_action_goal_handle_t * goal_handle)
+{
+  rclc_action_goal_handle_t * handle = *list;
+  if (goal_handle == handle) {
+    *list = handle->next;
+    return goal_handle;
+  }
+
+  while (NULL != handle) {
+    if (handle->next == goal_handle) {
+      handle->next = goal_handle->next;
+      return goal_handle;
+    }
+    handle = handle->next;
+  }
+  return NULL;
+}
+
+rclc_action_goal_handle_t * rclc_action_take_goal_handle(
+  void * untyped_entity)
+{
+  rclc_generic_entity_t * entity = (rclc_generic_entity_t *) untyped_entity;
+  rclc_action_goal_handle_t * handle = rclc_action_pop_goal_handle_from_list(
+    &entity->free_goal_handles);
+
+  if (NULL != handle) {
+    // Initialize handle
+    handle->available_goal_response = false;
+    handle->goal_accepted = false;
+    handle->available_feedback = false;
+    handle->available_result_response = false;
+    handle->available_cancel_response = false;
+    handle->goal_cancelled = false;
+    handle->status = GOAL_STATE_UNKNOWN;
+    rclc_action_put_goal_handle_in_list(&entity->used_goal_handles, handle);
+  }
+
+  return handle;
+}
+
+void rclc_action_init_goal_handle_memory(
+  void * untyped_entity)
+{
+  rclc_generic_entity_t * entity = (rclc_generic_entity_t *) untyped_entity;
+  entity->free_goal_handles = entity->goal_handles_memory;
+  size_t size = entity->goal_handles_memory_size;
+  for (size_t i = 0; i < size - 1; i++) {
+    entity->free_goal_handles[i].next = &entity->free_goal_handles[i + 1];
+  }
+  entity->free_goal_handles[size - 1].next = NULL;
+}
+
+void rclc_action_put_goal_handle(
+  void * untyped_entity,
+  rclc_action_goal_handle_t * goal_handle)
+{
+  rclc_generic_entity_t * entity = (rclc_generic_entity_t *) untyped_entity;
+  rclc_action_goal_handle_t * handle = rclc_action_take_goal_handle_from_list(
+    &entity->used_goal_handles, goal_handle);
+  if (NULL != handle) {
+    rclc_action_put_goal_handle_in_list(&entity->free_goal_handles, handle);
+  }
+}
+
+rclc_action_goal_handle_t * rclc_action_get_handle_by_uuid(
+  void * untyped_entity,
+  const unique_identifier_msgs__msg__UUID * uuid)
+{
+  rclc_generic_entity_t * entity = (rclc_generic_entity_t *) untyped_entity;
+  rclc_action_goal_handle_t * handle = entity->used_goal_handles;
+  while (NULL != handle) {
+    if (uuidcmp(handle->goal_id.uuid, uuid->uuid)) {
+      return handle;
+    }
+    handle = handle->next;
+  }
+  return handle;
+}
+
+rclc_action_goal_handle_t * rclc_action_get_first_handle_by_status(
```

Comment:
Also consider to replace `get` by `find` since it is not a simple getter, but a list search.

Reply:
1a3be5faabe5abe581648730004035dd4c4ff68e